cmake: Fix library ABI versioning

hebasto opened the pull request:
This change emulates Libtool to make sure Libtool and CMake agree on the ABI version.

To test, one needs to simulate a release with backward-compatible API changes, which means the following changes in `configure.ac` and `CMakeLists.txt`:
- incrementing of `*_LIB_VERSION_CURRENT`
- setting `*_LIB_VERSION_REVISION` to zero
- incrementing of `*_LIB_VERSION_AGE`

real-or-random commented:
Sorry, I lost context here a bit. Can you explain again what the issue is and how it's fixed?

hebasto commented:
> Sorry, I lost context here a bit. Can you explain again what the issue is and how it's fixed?

1. On the master branch (3bab71cf0534fd5948386de2b6ac31a7044ad76b), simulate a release with backward-compatible API change with the diff as follows:
```diff
diff --git a/CMakeLists.txt b/CMakeLists.txt
index a70165e..8216bad 100644
--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -17,9 +17,9 @@ project(libsecp256k1 VERSION 0.3.2 LANGUAGES C)
 # https://www.gnu.org/software/libtool/manual/html_node/Updating-version-info.html
 # All changes in experimental modules are treated as if they don't affect the
 # interface and therefore only increase the revision.
-set(${PROJECT_NAME}_LIB_VERSION_CURRENT 2)
-set(${PROJECT_NAME}_LIB_VERSION_REVISION 2)
-set(${PROJECT_NAME}_LIB_VERSION_AGE 0)
+set(${PROJECT_NAME}_LIB_VERSION_CURRENT 3)
+set(${PROJECT_NAME}_LIB_VERSION_REVISION 0)
+set(${PROJECT_NAME}_LIB_VERSION_AGE 1)
 
 set(CMAKE_C_STANDARD 90)
 set(CMAKE_C_EXTENSIONS OFF)
diff --git a/configure.ac b/configure.ac
index 0b555ea..acb923e 100644
--- a/configure.ac
+++ b/configure.ac
@@ -13,9 +13,9 @@ define(_PKG_VERSION_IS_RELEASE, false)
 # https://www.gnu.org/software/libtool/manual/html_node/Updating-version-info.html
 # All changes in experimental modules are treated as if they don't affect the
 # interface and therefore only increase the revision.
-define(_LIB_VERSION_CURRENT, 2)
-define(_LIB_VERSION_REVISION, 2)
-define(_LIB_VERSION_AGE, 0)
+define(_LIB_VERSION_CURRENT, 3)
+define(_LIB_VERSION_REVISION, 0)
+define(_LIB_VERSION_AGE, 1)
 
 AC_INIT([libsecp256k1],m4_join([.], _PKG_VERSION_MAJOR, _PKG_VERSION_MINOR, _PKG_VERSION_PATCH)m4_if(_PKG_VERSION_IS_RELEASE, [true], [], [-dev]),[https://github.com/bitcoin-core/secp256k1/issues],[libsecp256k1],[https://github.com/bitcoin-core/secp256k1])
 
```

2. Build using Autotools:
```
$ ./autogen.sh && ./configure && make
$ ls -1 .libs/libsecp256k1.so.*
.libs/libsecp256k1.so.2
.libs/libsecp256k1.so.2.1.0
```

3. Build using CMake:
```
$ cmake -S . -B build && cmake --build build
$ ls -1 build/src/libsecp256k1.so.*
build/src/libsecp256k1.so.3
build/src/libsecp256k1.so.3.1.0
```

hebasto commented:
> Have you checked this against libtool on MacOS too (including the `-current_version` and `-compatibility_version` passed to the linker)?

Checked. It doesn't look right. Investigating...

> Do we care about Windows?

I've noticed this issue while working on Windows stuff :)

hebasto commented:
> > Have you checked this against libtool on MacOS too (including the `-current_version` and `-compatibility_version` passed to the linker)?
> 
> Checked. It doesn't look right. Investigating...

To make versioning 100%-compatible with Libtool's scheme for macOS, [`MACHO_COMPATIBILITY_VERSION`](https://cmake.org/cmake/help/latest/prop_tgt/MACHO_COMPATIBILITY_VERSION.html) and [`MACHO_CURRENT_VERSION`](https://cmake.org/cmake/help/latest/prop_tgt/MACHO_CURRENT_VERSION.html) are required, but they are available in CMake 3.17+ only.

hebasto commented:
> Can you add a comment that says that this emulates libtool, something like,
> 
> ```
>  This emulates libtool to make sure libtool and cmake agree on the ABI version,
>  see below "Calculate the version variables" in build-aux/ltmain.sh.
> ```

Thanks! Added.

real-or-random commented:
> I've limited this PR to Linux platform explicitly, leaving other platforms discussion for follow ups.

Concept ACK, but I fail to see how the current diff affects Linux only


Given that you looked into this, what would you recommend for other problems? I think we could make it work on CMake 3.17+ and perhaps output a warning on lower versions (or document it somewhere else)? What about Windows?

hebasto commented:
> > I've limited this PR to Linux platform explicitly, leaving other platforms discussion for follow ups.
> 
> Concept ACK, but I fail to see how the current diff affects Linux only
> 
> Given that you looked into this, what would you recommend for other problems? I think we could make it work on CMake 3.17+ and perhaps output a warning on lower versions (or document it somewhere else)? What about Windows?

Reworked to be applied for all supported target platforms.

Added a warning if the user is going to build macOS DYLIB using CMake < 3.17.

real-or-random commented on src/CMakeLists.txt:
```diff
@@ -34,10 +34,44 @@ set_target_properties(secp256k1_precomputed PROPERTIES POSITION_INDEPENDENT_CODE
+  set_target_properties(secp256k1 PROPERTIES
+    VERSION ${${PROJECT_NAME}_soversion}.${${PROJECT_NAME}_LIB_VERSION_AGE}.${${PROJECT_NAME}_LIB_VERSION_REVISION}
+  )
+elseif(CMAKE_SYSTEM_NAME STREQUAL "Darwin")
```
```suggestion
elseif(APPLE)
```
Would this make a difference? (If not, I'm not sure if it's better ^^.)

hebasto replied:
Thanks! [Updated](https://github.com/bitcoin-core/secp256k1/pull/1270#issuecomment-1522427772).

hebasto commented:
Updated c6fd4263af266fd9ec52d6dda90d66e55ce79871 -> 5a730304c55d77f377072993f9c76688a9f19246 ([pr1270.04](https://github.com/hebasto/secp256k1/commits/pr1270.04) -> [pr1270.05](https://github.com/hebasto/secp256k1/commits/pr1270.05), [diff](https://github.com/hebasto/secp256k1/compare/pr1270.04..pr1270.05)):
- addressed @real-or-random's [comment](https://github.com/bitcoin-core/secp256k1/pull/1270#discussion_r1176687542)

theuni commented:
Concept ACK.

It's frustrating that they differ. Also, note that this is a case where a user gets _worse_ binaries with older CMake, but I'm not sure we can do anything about that.

real-or-random reviewed (approved):
ACK https://github.com/bitcoin-core/secp256k1/pull/1270/commits/5a730304c55d77f377072993f9c76688a9f19246 diff looks good and I tested on Linux

@hebasto How confident are you about macOS and Windows? If you've tested on these platforms, I think this is ready for merge.

hebasto commented:
> @hebasto How confident are you about macOS and Windows?

Tested once more with the following diff:
```diff
diff --git a/CMakeLists.txt b/CMakeLists.txt
index 91d2bed..22706d3 100644
--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -17,9 +17,9 @@ project(libsecp256k1 VERSION 0.3.2 LANGUAGES C)
 # https://www.gnu.org/software/libtool/manual/html_node/Updating-version-info.html
 # All changes in experimental modules are treated as if they don't affect the
 # interface and therefore only increase the revision.
-set(${PROJECT_NAME}_LIB_VERSION_CURRENT 2)
-set(${PROJECT_NAME}_LIB_VERSION_REVISION 2)
-set(${PROJECT_NAME}_LIB_VERSION_AGE 0)
+set(${PROJECT_NAME}_LIB_VERSION_CURRENT 1000)
+set(${PROJECT_NAME}_LIB_VERSION_REVISION 100)
+set(${PROJECT_NAME}_LIB_VERSION_AGE 10)
 
 set(CMAKE_C_STANDARD 90)
 set(CMAKE_C_EXTENSIONS OFF)
diff --git a/configure.ac b/configure.ac
index 0b555ea..08d9b64 100644
--- a/configure.ac
+++ b/configure.ac
@@ -13,9 +13,9 @@ define(_PKG_VERSION_IS_RELEASE, false)
 # https://www.gnu.org/software/libtool/manual/html_node/Updating-version-info.html
 # All changes in experimental modules are treated as if they don't affect the
 # interface and therefore only increase the revision.
-define(_LIB_VERSION_CURRENT, 2)
-define(_LIB_VERSION_REVISION, 2)
-define(_LIB_VERSION_AGE, 0)
+define(_LIB_VERSION_CURRENT, 1000)
+define(_LIB_VERSION_REVISION, 100)
+define(_LIB_VERSION_AGE, 10)
 
 AC_INIT([libsecp256k1],m4_join([.], _PKG_VERSION_MAJOR, _PKG_VERSION_MINOR, _PKG_VERSION_PATCH)m4_if(_PKG_VERSION_IS_RELEASE, [true], [], [-dev]),[https://github.com/bitcoin-core/secp256k1/issues],[libsecp256k1],[https://github.com/bitcoin-core/secp256k1])
 
```

On **macOS**:
- Autotools:
```
% ./autogen.sh && ./configure --disable-static
% make
% ls -l .libs/*dylib*
-rwxr-xr-x  1 hebasto  staff  1361344 Apr 28 19:58 .libs/libsecp256k1.990.dylib
lrwxr-xr-x  1 hebasto  staff       22 Apr 28 19:58 .libs/libsecp256k1.dylib -> libsecp256k1.990.dylib
% otool -L .libs/libsecp256k1.dylib 
.libs/libsecp256k1.dylib:
 /usr/local/lib/libsecp256k1.990.dylib (compatibility version 1001.0.0, current version 1001.100.0)
 /usr/lib/libSystem.B.dylib (compatibility version 1.0.0, current version 1319.0.0)
```
- CMake:
```
% ls -l build/src/*dylib*
-rwxr-xr-x  1 hebasto  staff  1361320 Apr 28 20:00 build/src/libsecp256k1.990.dylib
lrwxr-xr-x  1 hebasto  staff       22 Apr 28 20:00 build/src/libsecp256k1.dylib -> libsecp256k1.990.dylib
% otool -L build/src/libsecp256k1.dylib 
build/src/libsecp256k1.dylib:
 @rpath/libsecp256k1.990.dylib (compatibility version 1001.0.0, current version 1001.100.0)
 /usr/lib/libSystem.B.dylib (compatibility version 1.0.0, current version 1319.0.0)
 ```

A minor issue has been noted when cross-building for Windows. Going to address it shortly.

@real-or-random  
> I think this is ready for merge.

Let's fix and test the issue for Windows first :)

hebasto commented:
Updated 5a730304c55d77f377072993f9c76688a9f19246 -> d2e4e18a1ab7761282a24e90b1a77d10af3c2de8 ([pr1270.05](https://github.com/hebasto/secp256k1/commits/pr1270.05) -> [pr1270.06](https://github.com/hebasto/secp256k1/commits/pr1270.06), [diff](https://github.com/hebasto/secp256k1/compare/pr1270.05..pr1270.06)):
- fixed the issue when cross compiling for Windows

On Ubuntu 23.04 (with the diff from https://github.com/bitcoin-core/secp256k1/pull/1270#issuecomment-1527919996):
- Autotools:
```
$ ./autogen.sh && ./configure --host=x86_64-w64-mingw32 --disable-static
$ ./configure --disable-static --host=x86_64-w64-mingw32
$ make
$ ls -1 .libs/*dll*
.libs/libsecp256k1-990.dll
.libs/libsecp256k1.dll.a
```

- CMake:
```
$ cmake -S . -B build -DCMAKE_TOOLCHAIN_FILE=cmake/x86_64-w64-mingw32.toolchain.cmake
$ cmake --build build
$ ls -1 build/src/*dll*
build/src/libsecp256k1-990.dll
build/src/libsecp256k1.dll.a
```

Not able to test native builds on Windows right now...

hebasto commented:
Updated d2e4e18a1ab7761282a24e90b1a77d10af3c2de8 -> bef448f9af248dba016883401de07b431f3e686e ([pr1270.06](https://github.com/hebasto/secp256k1/commits/pr1270.06) -> [pr1270.07](https://github.com/hebasto/secp256k1/commits/pr1270.07), [diff](https://github.com/hebasto/secp256k1/compare/pr1270.06..pr1270.07)):
- fixed all Windows-related issues

@real-or-random 
> How confident are you about macOS and Windows?

100%

> If you've tested on these platforms, I think this is ready for merge.

I've tested cross-compiling for Windows and native builds on macOS and on Windows.

real-or-random commented on src/CMakeLists.txt:
```diff
@@ -34,10 +34,44 @@ set_target_properties(secp256k1_precomputed PROPERTIES POSITION_INDEPENDENT_CODE
+    )
+  endif()
+elseif(CMAKE_SYSTEM_NAME STREQUAL "Windows")
+  set(${PROJECT_NAME}_windows "secp256k1")
+  if(MSVC)
+    set(${PROJECT_NAME}_windows "${PROJECT_NAME}")
+  endif()
```
For my understanding: Why does this need to be different on MSVC?

hebasto replied:
MSVC skips adding a "lib" prefix by default. See the item 2 in https://github.com/bitcoin-core/secp256k1/pull/1237#issue-1619808558 for additional details.

hebasto commented on src/CMakeLists.txt:
```diff
@@ -34,10 +34,44 @@ set_target_properties(secp256k1_precomputed PROPERTIES POSITION_INDEPENDENT_CODE
+  endif()
+  set_target_properties(secp256k1 PROPERTIES
+    ARCHIVE_OUTPUT_NAME "${${PROJECT_NAME}_windows}"
+    RUNTIME_OUTPUT_NAME "${${PROJECT_NAME}_windows}-${${PROJECT_NAME}_soversion}"
```
CMake 3.27+ has [`DLL_NAME_WITH_SOVERSION`](https://cmake.org/cmake/help/latest/prop_tgt/DLL_NAME_WITH_SOVERSION.html) on the same purpose.
